cpp/src/arrow/filesystem/localfs.cc

```diff
@@ -46,6 +46,35 @@ using ::arrow::internal::IOErrorFromWinError;
 using ::arrow::internal::NativePathString;
 using ::arrow::internal::PlatformFilename;
 
+namespace internal {
+
+#ifdef _WIN32
+static bool IsDriveLetter(char c) {
+  // Can't use locale-dependent functions from the C/C++ stdlib
+  return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z');
+}
+#endif
+
+bool DetectAbsolutePath(const std::string& s) {
+  // Is it a /-prefixed local path?
+  if (s.length() >= 1 && s[0] == '/') {
+    return true;
+  }
+#ifdef _WIN32
+  // Is it a \-prefixed local path?
+  if (s.length() >= 1 && s[0] == '\\') {
+    return true;
+  }
+  // Does it start with a drive letter, e.g. "C:..."?
```

Comment:
`C:foo\bar` is a relative path; the path is not absolute unless the third character is a slash

Reply:
Yes, but that's intended here (see docstring).

Reply:
I don't understand. The name of the function is `DetectAbsolutePath`. In several places, it's commented that only an absolute path will be detected. The docstring says "local absolute path", but to me (in the context of potentially remote paths) that refers to "local to this filesystem".

To be clear, what I mean is:
```sh
C:\Users\bkietz> dir C:arrow > relative.list
C:\Users\bkietz> dir C:\Users\bkietz\arrow > absolute.list
C:\Users\bkietz> diff relative.list absolute.list
# empty
```

Reply:
I used the word "absolute" because it's, well, terse enough and commonly understood :-) Do you have a better name for "absolute or starts with a drive letter"?

(I'm not sure what your `diff` example is supposed to show, by the way)

Reply:
To me, absolute means "independent of current working directory", while relative means "the current working directory is an implicit prefix". The diff example was intended to show that when `C:arrow` is evaluated, the current working directory (`C:\Users\bkietz`) is prepended to it so `dir C:arrow` yields the same results as the absolute version `dir C:\Users\bkietz\arrow`.

Is it your intention that the current working directory be an implicit prefix when a drive letter (but no leading `\`) is present?

Reply:
> Is it your intention that the current working directory be an implicit prefix when a drive letter (but no leading `\`) is present?

Right. The decisive factor here is that such paths are not URIs (or at least they have no chance of resolving to an existing filesystem implementation).

Reply:
So IIUC in addition to absolute paths we'll accept relative paths on windows if they have a drive prefix, because they are unambiguously not URIs

It seems simpler to remove that caveat and require absolute paths on or off windows than to document it :)

I don't understand the added value of accepting this corner case of relative paths, but I won't block on this anymore

Reply:
Hmm, it does sound like a good idea to restrict to absolute paths. Will see on Monday :-)